### PR TITLE
[FAW] parallel faw embedding

### PR DESCRIPTION
**Pull request by feifeibear**
From https://github.com/zxgx/FreqCacheEmbedding/blob/main/recsys/modules/embeddings/freq_aware_embedding.py

### CODE REVIEW

File: colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        self.rank = torch.distributed.get_rank()
+        self.world_size = torch.distributed.get_world_size()
+        self.debug = debug
+
+        self.partition_start_index, self.partition_end_index, divisible = get_partition(
+            embedding_dim, self.rank, self.world_size)
+        self.embedding_dim_per_partition = self.partition_end_index - self.partition_start_index
+
+        if _weight is None:
+            self._weight.process_group = ProcessGroup(tp_degree=self.world_size)
```
Comment (zxgx) [inline colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py:60]: 这里能跑通吗？没有看到self._weight的初始化？

Comment (zxgx) [reply colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py:60]: 是不是顺序反了

File: colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        self._weight.data.uniform_(-1 / self.num_embeddings, 1 / self.num_embeddings)
+        if self.padding_idx is not None:
+            self._weight[self.padding_idx].fill_(0)
+
+    def preprocess(self,
+                   cuda_row_num: int,
+                   ids_freq_mapping: Optional[List[int]] = None,
+                   warmup_ratio: float = 0.7,
+                   buffer_size: int = 50_000):
+        self.cache_weight_mgr = CachedParamMgr(self._weight, cuda_row_num, buffer_size=buffer_size)
```
Comment (zxgx) [inline colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py:96]: 我一直觉得这个地方设计有问题，self._weight在CacheParamMgr里会被复制成cpu_weight，这里的_weight似乎永远没有被用到，实际被更新的参数也是CacheParamMgr.cpu_weight，这也是之前为什么cpu memory开销这么高的原因

File: colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py (hunk 1 above)
Comment (feifeibear) [reply colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py:60]: L61 initialize a weight from empty?
But, this branch has not been tested.

File: colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py (hunk 2 above)
Comment (feifeibear) [reply colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py:96]: Do you mean self._weight has never been used?
You can try to optimize it later.

**Review by zxgx — APPROVED** [review]

**Pull request closed without merging**